zeebe/engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/DeploymentReconstructProcessor.java

```diff
@@ -42,7 +53,134 @@ public DeploymentReconstructProcessor(
   @Override
   public void processRecord(final TypedRecord<DeploymentRecord> record) {
     final var key = keyGenerator.nextKey();
-    // TODO: Only append the event when there are no more deployments to reconstruct
-    stateWriter.appendFollowUpEvent(key, DeploymentIntent.RECONSTRUCTED_ALL, record.getValue());
+    final var resource = findNextResource();
+    if (resource == null) {
+      stateWriter.appendFollowUpEvent(key, DeploymentIntent.RECONSTRUCTED_ALL, record.getValue());
+      return;
+    }
+
+    final DeploymentRecord deploymentRecord;
+    if (resource.deploymentKey() != NO_DEPLOYMENT_KEY) {
+      final var allResourcesOfDeployment = findResourcesWithDeploymentKey(resource.deploymentKey());
+      deploymentRecord =
+          recreateDeploymentForResources(
+              resource.deploymentKey(), resource.tenantId(), allResourcesOfDeployment);
+    } else {
+      deploymentRecord = createNewDeploymentForResource(resource);
+    }
+    stateWriter.appendFollowUpEvent(key, DeploymentIntent.RECONSTRUCTED, deploymentRecord);
+  }
+
+  private Resource findNextResource() {
+    final var foundProcess = new MutableReference<PersistedProcess>();
+    processState.forEachProcess(
+        null, // TODO: Continue where we left off
+        process -> {
+          final var deploymentKey = process.getDeploymentKey();
+          if (deploymentKey != NO_DEPLOYMENT_KEY
+              && deploymentState.hasStoredDeploymentRecord(deploymentKey)) {
+            return true;
```

Comment:
💭 I wonder if we can actually stop here if we found a deployment key that already has a stored deployment associated with it. I think if this is the case it must've been created on the latest version where we store this, so all processes after would also have been created on that same version.

Reply:
> I think if this is the case it must've been created on the latest version where we store this

Or we already recreated it. I think this should be possible but depends on how exactly we determine the `previousProcess`. So we should tackle this together in a later PR.